usaon_vta_survey/routes/response/relationships/observing_system_data_product.py

```diff
@@ -0,0 +1,222 @@
+    '/response/<string:survey_id>/observing_system_data_product_relationships',
+    methods=['GET', 'POST'],
+)
+def view_response_observing_system_data_product_relationships(survey_id: str):
+    """View and add observing system/dataproduct relationships to a response.
+
+    TODO: Refactor this whole pile of stuff. Less string magic. Less cyclomatic
+    complexity.
+    """
+    data_product_id, observing_system_id = _request_args(request)
+    survey = db.get_or_404(Survey, survey_id)
+
+    class SuperForm(FlaskForm):
+        """Combine all necessary forms into one super-form.
+
+        NOTE: Additional class attributes are added dynamically below.
+        """
+
+        relationship = FormField(FORMS_BY_MODEL[ResponseObservingSystemDataProduct])
+
+    response_observing_system_data_product = _response_observing_system_data_product(
+        data_product_id=data_product_id,
+        observing_system_id=observing_system_id,
+    )
+
+    response_data_product = _response_data_product(
+        data_product_id=data_product_id,
+        response_id=survey.response_id,
+    )
+
+    response_observing_system = _response_observing_system(
+        observing_system_id=observing_system_id,
+        response_id=survey.response_id,
+    )
+
+    _update_super_form(
+        SuperForm,
+        observing_system_id=observing_system_id,
+        data_product_id=data_product_id,
+    )
+    _update_relationship(
+        response_observing_system_data_product,
+        observing_system_id=observing_system_id,
+        data_product_id=data_product_id,
+    )
+
+    form_obj: dict[
+        str,
+        ResponseObservingSystem
+        | ResponseDataProduct
+        | ResponseObservingSystemDataProduct,
+    ] = {
+        'observing_system': response_observing_system,
+        'data_product': response_data_product,
+        # NOTE: Logic below depends on relationship being last in this dict
+        'relationship': response_observing_system_data_product,
+    }
+
+    if request.method == 'POST':
+        limit_response_editors()
+        form = SuperForm(request.form, obj=form_obj)
+
+        if form.validate():
+            # Add only submitted sub-forms into the db session
+            for key, obj in form_obj.items():
+                if hasattr(form, key):
+                    form[key].form.populate_obj(obj)
+                    db.session.add(obj)
+
+                    # Update the relationship object with the ids of any new entities
+                    if type(obj) is not ResponseObservingSystemDataProduct:
+                        # Get the db object's new ID
+                        db.session.flush()
+                        db.session.refresh(obj)
+
+                        # Update the relationship db object
+                        setattr(
+                            response_observing_system_data_product,
+                            f'response_{key}_id',
+                            obj.id,
+                        )
```

Comment:
I'd love to extract this whole block to a re-usable function if we can! Maybe `update_response_object_relationship_from_super_form(model: Type[Model], form: Form) -> None`? :sweat_smile: 

Maybe for the next PR as well, just sharing thoughts! :)